components/raftstore-v2/src/worker/pd/slowness.rs

```diff
@@ -0,0 +1,149 @@
+use tikv_util::trend::{RequestPerSecRecorder, Trend};
+
+use super::Runner;
+pub struct SlownessStatistics {
+    /// Detector to detect NetIo&DiskIo jitters.
+    slow_cause: Trend,
+    /// Reactor as an assistant detector to detect the QPS jitters.
+    slow_result: Trend,
+    slow_result_recorder: RequestPerSecRecorder,
+    last_tick_id: u64,
+    last_tick_finished: bool,
+}
+
+impl SlownessStatistics {
+    #[inline]
+    pub fn new(cfg: &Config) -> Self {
+        Self {
+            slow_cause: Trend::new(
+                // Disable SpikeFilter for now
+                Duration::from_secs(0),
+                STORE_SLOW_TREND_MISC_GAUGE_VEC.with_label_values(&["spike_filter_value"]),
+                STORE_SLOW_TREND_MISC_GAUGE_VEC.with_label_values(&["spike_filter_count"]),
+                Duration::from_secs(180),
+                Duration::from_secs(30),
+                Duration::from_secs(120),
+                Duration::from_secs(600),
+                1,
+                tikv_util::time::duration_to_us(Duration::from_micros(500)),
+                STORE_SLOW_TREND_MARGIN_ERROR_WINDOW_GAP_GAUGE_VEC.with_label_values(&["L1"]),
+                STORE_SLOW_TREND_MARGIN_ERROR_WINDOW_GAP_GAUGE_VEC.with_label_values(&["L2"]),
+                cfg.slow_trend_unsensitive_cause,
+            ),
+            slow_result: Trend::new(
+                // Disable SpikeFilter for now
+                Duration::from_secs(0),
+                STORE_SLOW_TREND_RESULT_MISC_GAUGE_VEC.with_label_values(&["spike_filter_value"]),
+                STORE_SLOW_TREND_RESULT_MISC_GAUGE_VEC.with_label_values(&["spike_filter_count"]),
+                Duration::from_secs(120),
+                Duration::from_secs(15),
+                Duration::from_secs(60),
+                Duration::from_secs(300),
+                1,
+                2000,
+                STORE_SLOW_TREND_RESULT_MARGIN_ERROR_WINDOW_GAP_GAUGE_VEC
+                    .with_label_values(&["L1"]),
+                STORE_SLOW_TREND_RESULT_MARGIN_ERROR_WINDOW_GAP_GAUGE_VEC
+                    .with_label_values(&["L2"]),
+                cfg.slow_trend_unsensitive_result,
+            ),
+            slow_result_recorder: RequestPerSecRecorder::new(),
+            last_tick_id: 0,
+            last_tick_finished: true,
+        }
+    }
+}
+
+impl<EK, ER, T> Runner<EK, ER, T>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: PdClient + 'static,
+{
+    /// Record slowness periodically.
+    pub fn handle_update_slowness_stats(&mut self, _tick: u64, duration: RaftstoreDuration) {
+        self.slowness_stats.last_tick_finished = true;
+        // Record IO latency on Disk.
+        self.slowness_stats.slow_cause.record(
+            tikv_util::time::duration_to_us(duration.sum()),
+            Instant::now(),
+        );
+    }
+
+    pub fn handle_slowness_stats_tick(&mut self) {
+        // Record a fairly great value when tick
+        self.slowness_stats
```

Comment:
This is confusing to me, we should record a greate value only when (the inspecting request is) timeout, but current code will record 100ms at every tick.

I understand that this (data in slowness) won't be sent if the store heartbeat to pd is un-delayed, but the data set in slowness still be abnormal.

Reply:
Yep, i've noticed this point. Originally, I wanna use a periodical tick(100ms) to make it tolerable to minor io jitters, but it's really confusing to others.

I'm considering whether to supplement more annotations or remove it.